File: frontend/services/agents/strategy-agent.ts
/**
 * Strategy Agent - Combines CMO (Chief Marketing Officer) and CFO (Chief Financial Officer) capabilities
 * Handles market research, financial analysis, and business strategy
 */

import { BaseMessage, HumanMessage, AIMessage } from '@langchain/core/messages';
import { tool } from '@langchain/core/tools';
import { z } from 'zod';

import BaseAgent from './base-agent';
import {
  AgentState,
  AgentConfig,
  AgentResponse,
  AgentTool,
  TaskStatus,
  TaskType,
  StrategyAgentTools,
  MarketResearchTool,
  FinancialAnalysisTool,
  CompetitiveIntelligenceTool
} from './types';
import GraphMemoryService from '../../utils/supabase/graph-memory';

export interface MarketInsight {
  trend: string;
  confidence: number;
  source: string;
  timeframe: string;
  impact: 'high' | 'medium' | 'low';
  recommendations: string[];
}

export interface FinancialMetric {
  name: string;
  value: number;
  unit: string;
  target?: number;
  variance?: number;
  trend: 'increasing' | 'decreasing' | 'stable';
}

export interface CompetitiveAnalysis {
  competitors: Array<{
    name: string;
    strengths: string[];
    weaknesses: string[];
    marketShare?: number;
    revenue?: number;
  }>;
  marketPosition: string;
  opportunities: string[];
  threats: string[];
  recommendations: string[];
}

export interface BusinessRecommendation {
  category: 'marketing' | 'financial' | 'strategic' | 'operational';
  priority: 'high' | 'medium' | 'low';
  title: string;
  description: string;
  expectedImpact: string;
  resources: string[];
  timeline: string;
  risks: string[];
  kpis: string[];
}

export class StrategyAgent extends BaseAgent {
  private strategyTools: StrategyAgentTools;
  private businessContext: any;
  private marketDataCache: Map<string, any> = new Map();
  private financialDataCache: Map<string, any> = new Map();

  constructor(config: AgentConfig, userId: string) {
    super(config, userId);

    this.strategyTools = {
      marketResearch: {
        webSearch: true,
        socialMediaAnalysis: true,
        trendIdentification: true,
        audienceSegmentation: true,
      },
      financialAnalysis: {
        budgetAnalysis: true,
        revenueProjection: true,
        costOptimization: true,
        riskAssessment: true,
      },
      competitiveIntelligence: {
        competitorTracking: true,
        marketPositioning: true,
        swotAnalysis: true,
        opportunityIdentification: true,
      }
    };
  }

  /**
   * Initialize strategy agent with business context
   */
  async initialize(): Promise<void> {
    await super.initialize();

    // Load business context from graph memory
    await this.loadBusinessContext();

    console.log('Strategy Agent initialized successfully');
  }

  /**
   * Process strategy-related tasks
   */
  async process(state: AgentState): Promise<AgentResponse> {
    const startTime = Date.now();

    try {
      await this.updateMetrics('start');

      const latestMessage = state.messages[state.messages.length - 1];
      const taskType = this.determineTaskType(latestMessage.content as string);

      let response: AgentResponse;

      switch (taskType) {
        case 'market_research':
          response = await this.handleMarketResearch(state);
          break;
        case 'financial_analysis':
          response = await this.handleFinancialAnalysis(state);
          break;
        case 'strategic_planning':
          response = await this.handleStrategicPlanning(state);
          break;
        default:
          response = await this.handleGeneralStrategy(state);
          break;
      }

      // Store insights in graph memory
      await this.storeInsights(response, taskType);

      const responseTime = Date.now() - startTime;
      await this.updateMetrics('complete', responseTime, response.success);

      return response;

    } catch (error) {
      console.error('Strategy Agent processing error:', error);
      await this.updateMetrics('error');

      return {
        success: false,
        message: `Error in strategy analysis: ${error}`,
        requiresUserInput: true,
        userInputPrompt: 'The strategy analysis encountered an error. Would you like to provide more context or try a different approach?'
      };
    }
  }

  /**
   * Handle market research tasks
   */
  private async handleMarketResearch(state: AgentState): Promise<AgentResponse> {
    const request = state.messages[state.messages.length - 1].content as string;

    try {
      // Extract market research parameters
      const researchParams = await this.extractResearchParameters(request);

      // Perform market analysis
      const marketInsights = await this.performMarketAnalysis(researchParams);

      // Generate recommendations
      const recommendations = await this.generateMarketingRecommendations(marketInsights, researchParams);

      // Create comprehensive response
      const analysis = {
        summary: this.generateMarketSummary(marketInsights),
        insights: marketInsights,
        recommendations,
        marketSize: await this.estimateMarketSize(researchParams),
        targetAudience: await this.analyzeTargetAudience(researchParams),
        competitiveLandscape: await this.analyzeCompetitiveLandscape(researchParams),
      };

      return {
        success: true,
        data: analysis,
        message: 'Market research completed successfully',
        nextAction: {
          type: 'continue',
          target: 'ceo',
          payload: { analysisType: 'market_research', insights: marketInsights.length }
        }
      };

    } catch (error) {
      console.error('Market research error:', error);
      return {
        success: false,
        message: `Market research failed: ${error}`
      };
    }
  }

  /**
   * Handle financial analysis tasks
   */
  private async handleFinancialAnalysis(state: AgentState): Promise<AgentResponse> {
    const request = state.messages[state.messages.length - 1].content as string;

    try {
      // Extract financial parameters
      const financialParams = await this.extractFinancialParameters(request);

      // Perform financial analysis
      const financialMetrics = await this.performFinancialAnalysis(financialParams);

      // Generate financial recommendations
      const recommendations = await this.generateFinancialRecommendations(financialMetrics, financialParams);

      // Risk assessment
      const riskAssessment = await this.performRiskAssessment(financialMetrics, financialParams);

      const analysis = {
        summary: this.generateFinancialSummary(financialMetrics),
        metrics: financialMetrics,
        recommendations,
        riskAssessment,
        projections: await this.generateFinancialProjections(financialMetrics),
        benchmarks: await this.getBenchmarkData(financialParams),
      };

      return {
        success: true,
        data: analysis,
        message: 'Financial analysis completed successfully',
        nextAction: {
          type: 'continue',
          target: 'ceo',
          payload: { analysisType: 'financial_analysis', metricsCount: financialMetrics.length }
        }
      };

    } catch (error) {
      console.error('Financial analysis error:', error);
      return {
        success: false,
        message: `Financial analysis failed: ${error}`
      };
    }
  }

  /**
   * Handle strategic planning tasks
   */
  private async handleStrategicPlanning(state: AgentState): Promise<AgentResponse> {
    const request = state.messages[state.messages.length - 1].content as string;

    try {
      // Extract strategic parameters
      const strategicParams = await this.extractStrategicParameters(request);

      // SWOT analysis
      const swotAnalysis = await this.performSWOTAnalysis(strategicParams);

      // Market positioning
      const marketPositioning = await this.analyzeMarketPositioning(strategicParams);

      // Growth opportunities
      const growthOpportunities = await this.identifyGrowthOpportunities(strategicParams);

      // Strategic roadmap
      const roadmap = await this.generateStrategicRoadmap(swotAnalysis, growthOpportunities, strategicParams);

      const plan = {
        executiveSummary: this.generateExecutiveSummary(swotAnalysis, growthOpportunities),
        swotAnalysis,
        marketPositioning,
        growthOpportunities,
        strategicInitiatives: roadmap.initiatives,
        timeline: roadmap.timeline,
        resourceRequirements: roadmap.resources,
        successMetrics: roadmap.kpis,
      };

      return {
        success: true,
        data: plan,
        message: 'Strategic plan developed successfully',
        nextAction: {
          type: 'continue',
          target: 'ceo',
          payload: { analysisType: 'strategic_planning', initiativesCount: roadmap.initiatives.length }
        }
      };

    } catch (error) {
      console.error('Strategic planning error:', error);
      return {
        success: false,
        message: `Strategic planning failed: ${error}`
      };
    }
  }

  /**
   * Handle general strategy tasks
   */
  private async handleGeneralStrategy(state: AgentState): Promise<AgentResponse> {
    const request = state.messages[state.messages.length - 1].content as string;

    try {
      // Use LLM with strategy-specific tools
      const tools = this.getStrategySpecificTools().map(t =>
        tool(t.handler, {
          name: t.name,
          description: t.description,
          schema: t.schema,
        })
      );

      const analysisPrompt = `As a Strategy Agent (CMO/CFO), analyze this business request:

Request: "${request}"

Consider:
- Market trends and competitive landscape
- Financial implications and ROI
- Strategic alignment with business goals
- Risk factors and mitigation strategies
- Implementation timeline and resources

Provide comprehensive analysis and actionable recommendations.`;

      const response = await this.llm.invoke([
        new HumanMessage(analysisPrompt),
        ...tools.map(t => ({ type: 'tool', tool: t.name } as any))
      ]);

      return {
        success: true,
        data: response.content,
        message: 'Strategy analysis completed'
      };

    } catch (error) {
      console.error('General strategy error:', error);
      return {
        success: false,
        message: `Strategy analysis failed: ${error}`
      };
    }
  }

  /**
   * Get strategy-specific tools
   */
  protected getStrategySpecificTools(): AgentTool[] {
    return [
      {
        name: 'market_trend_analysis',
        description: 'Analyze current market trends and provide insights',
        schema: z.object({
          industry: z.string().describe('Industry to analyze'),
          timeframe: z.string().optional().describe('Timeframe for analysis (e.g., 6 months, 1 year)'),
          region: z.string().optional().describe('Geographic region'),
        }),
        handler: async (input) => {
          try {
            const trends = await this.performMarketTrendAnalysis(input);
            return JSON.stringify(trends, null, 2);
          } catch (error) {
            return `Error analyzing market trends: ${error}`;
          }
        }
      },
      {
        name: 'competitor_analysis',
        description: 'Analyze competitors and market positioning',
        schema: z.object({
          competitors: z.array(z.string()).describe('List of competitors to analyze'),
          focusArea: z.string().optional().describe('Specific area to focus on (e.g., pricing, features, market share)'),
        }),
        handler: async (input) => {
          try {
            const analysis = await this.performCompetitorAnalysis(input);
            return JSON.stringify(analysis, null, 2);
          } catch (error) {
            return `Error analyzing competitors: ${error}`;
          }
        }
      },
      {
        name: 'financial_modeling',
        description: 'Create financial models and projections',
        schema: z.object({
          modelType: z.enum(['revenue', 'cost', 'profitability', 'valuation']).describe('Type of financial model'),
          parameters: z.record(z.any()).describe('Model parameters and assumptions'),
          timeframe: z.string().describe('Projection timeframe'),
        }),
        handler: async (input) => {
          try {
            const model = await this.createFinancialModel(input);
            return JSON.stringify(model, null, 2);
          } catch (error) {
            return `Error creating financial model: ${error}`;
          }
        }
      },
      {
        name: 'customer_segmentation',
        description: 'Analyze and segment customer bases',
        schema: z.object({
          data: z.any().describe('Customer data for segmentation'),
          segments: z.number().optional().describe('Target number of segments'),
          criteria: z.array(z.string()).optional().describe('Segmentation criteria'),
        }),
        handler: async (input) => {
          try {
            const segmentation = await this.performCustomerSegmentation(input);
            return JSON.stringify(segmentation, null, 2);
          } catch (error) {
            return `Error performing customer segmentation: ${error}`;
          }
        }
      },
      ...this.config.tools
    ];
  }

  /**
   * Market analysis methods
   */
  private async performMarketTrendAnalysis(params: any): Promise<MarketInsight[]> {
    // Check cache first
    const cacheKey = JSON.stringify(params);
    if (this.marketDataCache.has(cacheKey)) {
      return this.marketDataCache.get(cacheKey);
    }

    // Use LLM to generate trend analysis based on current knowledge
    const prompt = `Analyze market trends for ${params.industry} industry${params.region ? ` in ${params.region}` : ''}.

Provide 5-7 key trends with:
- Trend description
- Confidence level (0-1)
- Timeframe
- Impact level
- Actionable recommendations

Consider technological, economic, social, and regulatory factors.`;

    const response = await this.llm.invoke(prompt);
    const trends = this.parseMarketInsights(response.content as string);

    // Cache results
    this.marketDataCache.set(cacheKey, trends);

    return trends;
  }

  private async performCompetitorAnalysis(params: any): Promise<CompetitiveAnalysis> {
    const prompt = `Analyze these competitors: ${params.competitors.join(', ')}.

Provide comprehensive analysis including:
- Strengths and weaknesses for each competitor
- Market share estimates
- Pricing strategies
- Product offerings
- Market positioning
- Opportunities and threats
- Strategic recommendations

Focus on ${params.focusArea || 'overall competitive landscape'}.`;

    const response = await this.llm.invoke(prompt);
    return this.parseCompetitiveAnalysis(response.content as string);
  }

  /**
   * Financial analysis methods
   */
  private async createFinancialModel(params: any): Promise<any> {
    const { modelType, parameters, timeframe } = params;

    let prompt = `Create a ${modelType} financial model for ${timeframe}.

Parameters: ${JSON.stringify(parameters, null, 2)}`;

    switch (modelType) {
      case 'revenue':
        prompt += `
Include:
- Revenue streams
- Growth rates
- Market penetration
- Customer acquisition
- Pricing strategy`;
        break;
      case 'cost':
        prompt += `
Include:
- Fixed costs
- Variable costs
- Operating expenses
- Capital expenditures
- Cost optimization opportunities`;
        break;
      case 'profitability':
        prompt += `
Include:
- Revenue projections
- Cost structure
- Profit margins
- Break-even analysis
- ROI calculations`;
        break;
      case 'valuation':
        prompt += `
Include:
- Discounted cash flow
- Comparable company analysis
- Market multiples
- Sensitivity analysis
- Valuation range`;
        break;
    }

    const response = await this.llm.invoke(prompt);
    return this.parseFinancialModel(response.content as string);
  }

  /**
   * Helper methods for parsing and analysis
   */
  private parseMarketInsights(text: string): MarketInsight[] {
    // Simple parsing - in production, would use more sophisticated JSON extraction
    const insights: MarketInsight[] = [];
    const lines = text.split('\n').filter(line => line.trim());

    let currentInsight: any = {};
    for (const line of lines) {
      if (line.includes('Trend:')) {
        if (Object.keys(currentInsight).length > 0) {
          insights.push(currentInsight as MarketInsight);
        }
        currentInsight = { trend: line.split('Trend:')[1].trim() };
      } else if (line.includes('Confidence:')) {
        currentInsight.confidence = parseFloat(line.split('Confidence:')[1].trim());
      } else if (line.includes('Impact:')) {
        currentInsight.impact = line.split('Impact:')[1].trim().toLowerCase() as any;
      }
    }

    if (Object.keys(currentInsight).length > 0) {
      insights.push(currentInsight as MarketInsight);
    }

    return insights;
  }

  private parseCompetitiveAnalysis(text: string): CompetitiveAnalysis {
    // Simplified parsing - would use structured extraction in production
    return {
      competitors: [],
      marketPosition: 'Analyzing...',
      opportunities: [],
      threats: [],
      recommendations: []
    };
  }

  private parseFinancialModel(text: string): any {
    // Simplified parsing - would use structured extraction in production
    return {
      summary: text.substring(0, 500),
      assumptions: {},
      projections: {},
      calculations: {}
    };
  }

  /**
   * Parameter extraction methods
   */
  private async extractResearchParameters(request: string): Promise<any> {
    const prompt = `Extract market research parameters from this request:

"${request}"

Return JSON with: industry, targetMarket, geographicScope, timeframe, competitors (array), focusAreas (array)`;

    const response = await this.llm.invoke(prompt);
    try {
      const jsonMatch = response.content as string;
      return JSON.parse(jsonMatch);
    } catch {
      return { industry: '', targetMarket: '', geographicScope: 'global', timeframe: '12 months' };
    }
  }

  private async extractFinancialParameters(request: string): Promise<any> {
    const prompt = `Extract financial analysis parameters from this request:

"${request}"

Return JSON with: analysisType, timeframe, currency, metrics (array), assumptions (object)`;

    const response = await this.llm.invoke(prompt);
    try {
      const jsonMatch = response.content as string;
      return JSON.parse(jsonMatch);
    } catch {
      return { analysisType: 'general', timeframe: '12 months', currency: 'USD' };
    }
  }

  private async extractStrategicParameters(request: string): Promise<any> {
    const prompt = `Extract strategic planning parameters from this request:

"${request}"

Return JSON with: businessGoals, timeframe, resources, constraints, focusAreas`;

    const response = await this.llm.invoke(prompt);
    try {
      const jsonMatch = response.content as string;
      return JSON.parse(jsonMatch);
    } catch {
      return { businessGoals: [], timeframe: '12 months', resources: [], constraints: [] };
    }
  }

  /**
   * Load and store business context
   */
  private async loadBusinessContext(): Promise<void> {
    try {
      const businessNodes = await this.graphMemory.getNodesByType('business_rule');
      this.businessContext = {
        industry: 'Technology',
        businessModel: 'SaaS',
        targetMarket: 'Small to Medium Businesses',
        recentTrends: businessNodes.map(node => node.content)
      };
    } catch (error) {
      console.warn('Could not load business context:', error);
      this.businessContext = {};
    }
  }

  private async storeInsights(response: AgentResponse, taskType: TaskType): Promise<void> {
    if (response.success && response.data) {
      try {
        await this.graphMemory.addMemory(
          JSON.stringify(response.data),
          'feature',
          [],
          { taskType, agent: 'strategy', timestamp: new Date().toISOString() }
        );
      } catch (error) {
        console.warn('Could not store insights:', error);
      }
    }
  }

  /**
   * Determine task type from content
   */
  private determineTaskType(content: string): TaskType {
    const contentLower = content.toLowerCase();

    if (contentLower.includes('market') || contentLower.includes('research') || contentLower.includes('competition')) {
      return 'market_research';
    }
    if (contentLower.includes('financial') || contentLower.includes('budget') || contentLower.includes('revenue') || contentLower.includes('cost')) {
      return 'financial_analysis';
    }
    if (contentLower.includes('strategy') || contentLower.includes('planning') || contentLower.includes('roadmap')) {
      return 'strategic_planning';
    }

    return 'strategic_planning';
  }

  /**
   * Placeholder methods for full implementation
   */
  private async performMarketAnalysis(params: any): Promise<any[]> { return []; }
  private async generateMarketingRecommendations(insights: any[], params: any): Promise<any[]> { return []; }
  private async estimateMarketSize(params: any): Promise<any> { return {}; }
  private async analyzeTargetAudience(params: any): Promise<any> { return {}; }
  private async analyzeCompetitiveLandscape(params: any): Promise<any> { return {}; }
  private generateMarketSummary(insights: any[]): string { return 'Market analysis summary'; }
  private async performFinancialAnalysis(params: any): Promise<FinancialMetric[]> { return []; }
  private async generateFinancialRecommendations(metrics: FinancialMetric[], params: any): Promise<any[]> { return []; }
  private async performRiskAssessment(metrics: FinancialMetric[], params: any): Promise<any> { return {}; }
  private generateFinancialSummary(metrics: FinancialMetric[]): string { return 'Financial analysis summary'; }
  private async generateFinancialProjections(metrics: FinancialMetric[]): Promise<any> { return {}; }
  private async getBenchmarkData(params: any): Promise<any> { return {}; }
  private async performSWOTAnalysis(params: any): Promise<any> { return {}; }
  private async analyzeMarketPositioning(params: any): Promise<any> { return {}; }
  private async identifyGrowthOpportunities(params: any): Promise<any[]> { return []; }
  private async generateStrategicRoadmap(swot: any, opportunities: any[], params: any): Promise<any> { return {}; }
  private generateExecutiveSummary(swot: any, opportunities: any[]): string { return 'Strategic plan executive summary'; }
  private async performCustomerSegmentation(params: any): Promise<any> { return {}; }
}

export default StrategyAgent;

File: backend/agents/swarm/strategy_agent.py
"""
Strategy Agent - Combined CMO/CFO agent for market analysis and financial planning
Handles business strategy, competitive analysis, market research, and financial planning
"""

import logging
from typing import Dict, Any, List, Optional
from datetime import datetime
import json

from langchain_core.messages import HumanMessage, AIMessage, SystemMessage
from langchain_openai import ChatOpenAI
from tavily import TavilyClient

logger = logging.getLogger(__name__)


class StrategyAgent:
    """Strategy Agent combining CMO and CFO responsibilities"""

    def __init__(self, config: Dict[str, Any]):
        self.config = config
        self.name = "Strategy Agent (CMO/CFO)"
        self.logger = logging.getLogger(f"{__name__}.{self.__class__.__name__}")

        # Initialize LLM
        self.llm = ChatOpenAI(
            model=config.get("model", "gpt-4"),
            temperature=config.get("temperature", 0.2),
            max_tokens=config.get("max_tokens", 2000),
            api_key=config.get("openai_api_key"),
        )

        # Initialize Tavily for web search
        self.tavily_client = TavilyClient(
            api_key=config.get("tavily_api_key")
        ) if config.get("tavily_api_key") else None

        # Strategy agent persona
        self.system_prompt = """
        You are the Strategy Agent for AutoAdmin, combining the expertise of a CMO and CFO.

        Your core responsibilities:

        AS CMO (Chief Marketing Officer):
        1. Market research and competitive analysis
        2. Brand strategy and positioning
        3. Customer acquisition and growth strategies
        4. Market trend identification
        5. Go-to-market strategy development

        AS CFO (Chief Financial Officer):
        1. Financial planning and analysis
        2. Budget allocation and optimization
        3. ROI analysis and investment decisions
        4. Risk assessment and mitigation
        5. Financial modeling and forecasting

        Your approach:
        - Use data-driven analysis and insights
        - Consider both market opportunities and financial implications
        - Provide actionable strategic recommendations
        - Balance growth objectives with financial responsibility
        - Focus on sustainable business development

        Always respond with:
        1. Strategic analysis and insights
        2. Data-driven recommendations
        3. Financial implications and ROI considerations
        4. Risk assessment and mitigation strategies
        5. Actionable next steps and timelines
        """

        self.logger.info("Strategy Agent initialized")

    async def process_task(self, task_input: Dict[str, Any]) -> Dict[str, Any]:
        """Process a strategy-related task"""
        try:
            messages = task_input.get("messages", [])
            ceo_guidance = task_input.get("ceo_guidance", {})
            task_analysis = task_input.get("task_analysis", {})

            # Determine task type and gather data
            task_type = self._classify_strategy_task(messages)
            research_data = await self._gather_research_data(messages, task_type)

            # Build strategy prompt
            prompt = self._build_strategy_prompt(messages, ceo_guidance, task_analysis, research_data)

            # Get strategy response
            response = await self.llm.ainvoke([
                SystemMessage(content=self.system_prompt),
                HumanMessage(content=prompt)
            ])

            # Parse and structure the response
            strategy_result = self._parse_strategy_response(response.content, task_type)

            return {
                "agent": "strategy",
                "task_type": task_type,
                "analysis": strategy_result,
                "research_data": research_data,
                "timestamp": datetime.now().isoformat(),
            }

        except Exception as e:
            self.logger.error(f"Error processing strategy task: {e}")
            return {
                "agent": "strategy",
                "error": str(e),
                "task_type": "unknown",
                "analysis": {},
            }

    def _classify_strategy_task(self, messages: List) -> str:
        """Classify the type of strategy task"""
        if not messages:
            return "general_strategy"

        content = messages[-1].content.lower()

        # CMO-related keywords
        cmo_keywords = [
            "market", "customer", "competition", "brand", "marketing",
            "growth", "acquisition", "positioning", "campaign", "segmentation"
        ]

        # CFO-related keywords
        cfo_keywords = [
            "financial", "budget", "investment", "roi", "profit", "cost",
            "revenue", "funding", "valuation", "forecast", "analysis"
        ]

        cmo_score = sum(1 for keyword in cmo_keywords if keyword in content)
        cfo_score = sum(1 for keyword in cfo_keywords if keyword in content)

        if cmo_score > cfo_score:
            return "marketing_strategy"
        elif cfo_score > cmo_score:
            return "financial_strategy"
        else:
            return "business_strategy"

    async def _gather_research_data(self, messages: List, task_type: str) -> Dict[str, Any]:
        """Gather relevant research data using web search"""
        research_data = {
            "web_search_results": [],
            "market_data": {},
            "competitor_analysis": {},
        }

        if not self.tavily_client:
            return research_data

        try:
            # Extract key terms for search
            content = messages[-1].content if messages else ""
            search_query = self._extract_search_query(content, task_type)

            if search_query:
                # Perform web search
                search_results = self.tavily_client.search(
                    query=search_query,
                    search_depth="advanced",
                    max_results=5,
                    include_domains=None,
                    exclude_domains=None
                )

                research_data["web_search_results"] = search_results.get("results", [])

            self.logger.info(f"Gathered {len(research_data['web_search_results'])} research results")

        except Exception as e:
            self.logger.error(f"Error gathering research data: {e}")

        return research_data

    def _extract_search_query(self, content: str, task_type: str) -> str:
        """Extract relevant search query from content"""
        # Simple extraction - in production, use more sophisticated NLP
        content_lower = content.lower()

        if "market research" in content_lower:
            return "market trends industry analysis 2024"
        elif "competitive analysis" in content_lower:
            return "competitive landscape market share analysis"
        elif "financial planning" in content_lower:
            return "financial planning business forecast 2024"
        elif "investment analysis" in content_lower:
            return "investment opportunities ROI analysis"
        else:
            # Extract key terms from the content
            words = content_lower.split()
            key_terms = [word for word in words if len(word) > 3][:5]
            return " ".join(key_terms) if key_terms else "business strategy analysis"

    def _build_strategy_prompt(self, messages: List, ceo_guidance: Dict, task_analysis: Dict, research_data: Dict) -> str:
        """Build the strategy analysis prompt"""
        latest_message = messages[-1].content if messages else ""

        prompt = f"""
        STRATEGIC ANALYSIS REQUEST

        Task: {latest_message}

        CEO Guidance: {ceo_guidance.get('full_response', 'No specific guidance provided')}

        Task Analysis: {task_analysis}

        Research Data Available: {len(research_data.get('web_search_results', []))} web sources found

        As Strategy Agent (CMO/CFO), please provide:

        1. MARKET ANALYSIS (CMO):
           - Market size and trends
           - Competitive landscape
           - Target audience insights
           - Market opportunities and threats

        2. FINANCIAL ANALYSIS (CFO):
           - Financial implications and requirements
           - ROI analysis and projections
           - Budget recommendations
           - Financial risk assessment

        3. STRATEGIC RECOMMENDATIONS:
           - Actionable strategic initiatives
           - Implementation timeline
           - Resource requirements
           - Success metrics and KPIs

        4. RISK MITIGATION:
           - Potential risks and challenges
           - Mitigation strategies
           - Contingency planning

        Please provide specific, data-driven recommendations that align with business objectives.
        """

        return prompt

    def _parse_strategy_response(self, response_content: str, task_type: str) -> Dict[str, Any]:
        """Parse and structure the strategy response"""
        analysis = {
            "market_analysis": {},
            "financial_analysis": {},
            "recommendations": [],
            "risk_assessment": [],
            "implementation_plan": {},
            "success_metrics": [],
        }

        # Simple parsing - in production, use more sophisticated NLP
        content_lower = response_content.lower()

        # Extract key sections based on patterns
        sections = {
            "market": "market_analysis",
            "financial": "financial_analysis",
            "recommend": "recommendations",
            "risk": "risk_assessment",
            "implement": "implementation_plan",
            "metric": "success_metrics",
        }

        # For now, store the full response and basic classification
        analysis["full_analysis"] = response_content
        analysis["task_type"] = task_type
        analysis["key_insights"] = self._extract_key_insights(response_content)

        return analysis

    def _extract_key_insights(self, content: str) -> List[str]:
        """Extract key insights from the response"""
        insights = []

        # Look for bullet points or numbered lists
        lines = content.split('\n')
        for line in lines:
            line = line.strip()
            # Look for bullet points, numbered lists, or key phrases
            if (line.startswith(('•', '-', '*', '1.', '2.', '3.', '4.', '5.')) or
                any(keyword in line.lower() for keyword in [
                    'key insight', 'important', 'critical', 'essential',
                    'recommend', 'should', 'must', 'consider'
                ])):
                insights.append(line)

        return insights[:10]  # Return top 10 insights

    async def get_market_analysis(self, query: str) -> Dict[str, Any]:
        """Get specific market analysis for a query"""
        try:
            if not self.tavily_client:
                return {"error": "Web search not available"}

            search_results = self.tavily_client.search(
                query=f"market analysis {query}",
                max_results=5,
                search_depth="advanced"
            )

            return {
                "query": query,
                "results": search_results.get("results", []),
                "timestamp": datetime.now().isoformat(),
            }

        except Exception as e:
            self.logger.error(f"Error in market analysis: {e}")
            return {"error": str(e)}

    async def get_financial_analysis(self, data: Dict[str, Any]) -> Dict[str, Any]:
        """Analyze financial data and provide insights"""
        try:
            prompt = f"""
            FINANCIAL ANALYSIS REQUEST

            Financial Data: {json.dumps(data, indent=2)}

            As CFO, please analyze this data and provide:
            1. Financial health assessment
            2. Key performance indicators
            3. Trends and patterns
            4. Risk factors
            5. Recommendations for improvement
            """

            response = await self.llm.ainvoke([
                SystemMessage(content=self.system_prompt),
                HumanMessage(content=prompt)
            ])

            return {
                "analysis": response.content,
                "data_summary": {
                    "metrics_provided": list(data.keys()) if isinstance(data, dict) else [],
                    "analysis_timestamp": datetime.now().isoformat(),
                }
            }

        except Exception as e:
            self.logger.error(f"Error in financial analysis: {e}")
            return {"error": str(e)}

    async def health_check(self) -> Dict[str, Any]:
        """Health check for Strategy agent"""
        return {
            "status": "healthy",
            "agent": "strategy",
            "capabilities": [
                "market_research",
                "competitive_analysis",
                "financial_planning",
                "investment_analysis",
                "budget_optimization",
                "risk_assessment"
            ],
            "services": {
                "web_search": self.tavily_client is not None,
                "llm_ready": True,
            },
            "timestamp": datetime.now().isoformat(),
        }